### PR TITLE
Issue 430: Exposing Segmentstore Service on single IP

### LINKED ISSUE #430: Exposing Segmentstore Service on single IP address and Different ports
### Description
Exposing Segmentstore Service on single IP address and Different ports

### Importance
Nice_to_have

### Location
this requires changes in pravega.go and segmentstore.go files

### PR DESCRIPTION
**Pull request by Prabhaker24**
### Change log description

Exposing Segmentstore Service on single IP address and Different ports

### Purpose of the change
fixes #430 

### What the code does
Currently, Segment store pods are exposed using external IP address which is given by Service, Each pod has its own service which has a unique IP address which can be used to communicate to the Segment store, after this change we can expose segment store services on the same I/P address and different port

### How to verify it
1) To verify this change we need Bare metal kubernetes cluster and we need to specify the below fields for it to work in 
   the manifest 
```
pravega:
    . . .
    segmentStoreLoadBalancerIP: "10.243.39.103"
    segmentStoreExternalTrafficPolicy: "cluster"
    segmentStoreSvcAnnotations:
      metallb.universe.tf/allow-shared-ip: "shared-ss-ip"
```
and after deploying pravega we should see that all the Segment Store services should be having same LoadBalancerIP and a different port.

2) In case of an environment which does not allow metallb service to provision the same LoadBalancerIP to services like PKS we should be able to run segment store services on different I/P addresses and same port.

I have verified both Scenarios work

### CODE REVIEW

File: pkg/controller/pravega/pravega_segmentstore.go [hunk 1]
```diff
@@ -491,6 +487,15 @@ func MakeSegmentStoreExternalServices(p *api.PravegaCluster) []*corev1.Service {
 				},
 			},
 		}
+		if strings.EqualFold(p.Spec.Pravega.SegmentStoreExternalTrafficPolicy, "cluster") == true {
```
Comment (gaddamas) [inline pkg/controller/pravega/pravega_segmentstore.go:490]: change "cluster" to "Cluster" to be consistent with k8s documentation

File: README.md [hunk 2]
```diff
@@ -143,6 +143,8 @@ http://<pravega-name>-pravega-controller.<namespace>:10080/
 
 Check out the [external access documentation](doc/external-access.md) if your clients need to connect to Pravega from outside Kubernetes.
 
+Check out the [exposing Segmentstore service on single I/P address ](https://github.com/pravega/pravega-operator/blob/4aa88641c3d5a1d5afbb2b9e628846639fd13290/doc/external-access.md#exposing-segmentstore-service-on-single-ip-address-and-different-ports) if your clients need to connect to Pravega Segment store on the same I/P address from outside Kubernetes.
```
Comment (bourgeoisor) [inline README.md:146]: Isn't `IP address` the standard way of writing it? (rather than `I/P`)

File: charts/pravega/templates/pravega.yaml [hunk 3]
```diff
@@ -33,6 +33,10 @@ spec:
     {{- if .Values.externalAccess.enabled }}
     controllerServiceAccountName: {{ .Values.serviceAccount.name }}
     segmentStoreServiceAccountName: {{ .Values.serviceAccount.name }}
+    {{- if .Values.segmentStore.service.segmentStoreLoadBalancerIP }}
```
Comment (bourgeoisor) [inline charts/pravega/templates/pravega.yaml:36]: Is that conditional needed at all? If we remove the conditional and `.Values.segmentStore.service.segmentStoreLoadBalancerIP` is null / not set, then it's just going to set it as below:
```
segmentStoreLoadBalancerIP: null`
```
Which is the same as not setting it (thus, we don't need that conditional)

File: doc/external-access.md [hunk 4]
```diff
@@ -238,3 +238,31 @@ IP:                       10.100.200.183
 LoadBalancer Ingress:     10.247.108.104
 . . .
 ```
+# Exposing Segmentstore Service on single IP address and Different ports
+
+For Exposing SegmentStoreservices on the same I/P address we will use MetalLB,
```
Comment (bourgeoisor) [inline doc/external-access.md:243]: Same comments on `IP address` vs `I/P address`

File: charts/pravega/README.md [hunk 5]
```diff
@@ -70,6 +70,8 @@ The following table lists the configurable parameters of the Pravega chart and t
 | `segmentStore.resources.limits.memory` | Memory limits for segmentStore | `4Gi` |
 | `segmentStore.service.type` | Override the segmentStore service type, if external access is enabled (LoadBalancer/NodePort) | |
 | `segmentStore.service.annotations` | Annotations to add to the segmentStore service, if external access is enabled | `{}` |
+| `segmentStore.service.segmentStoreLoadBalancerIP` |It is used to provide a Load LoadBalancerIP | |
```
Comment (SrishT) [inline charts/pravega/README.md:73]: it says `Load LoadBalancerIP`

File: charts/pravega/templates/pravega.yaml [hunk 6]
```diff
@@ -6,7 +6,7 @@ metadata:
   labels:
 {{ include "pravega.commonLabels" . | indent 4 }}
 spec:
-  {{- if .Values.tls }}
+ {{- if .Values.tls }}
```
Comment (SrishT) [inline charts/pravega/templates/pravega.yaml:9]: Any specific reason for changing the indentation here?

File: charts/pravega/templates/pravega.yaml [hunk 7]
```diff
@@ -45,7 +49,7 @@ spec:
     {{- end }}
     {{- if .Values.segmentStore.service.annotations }}
     segmentStoreSvcAnnotations:
-{{ toYaml .Values.segmentStore.service.annotations | indent 6 }}
+{{ toYaml .Values.segmentStore.service.annotations | indent 6}}
```
Comment (SrishT) [inline charts/pravega/templates/pravega.yaml:52]: could u revert the change here?

File: pkg/apis/pravega/v1beta1/pravega.go [hunk 8]
```diff
@@ -147,6 +147,12 @@ type PravegaSpec struct {
 
 	// Annotations to be added to the external service
 	SegmentStoreServiceAnnotations map[string]string `json:"segmentStoreSvcAnnotations"`
+
+	// Specifying this I/P would insure we use same I/P address for all the ss services
```
Comment (SrishT) [inline pkg/apis/pravega/v1beta1/pravega.go:151]: it should be `ensure`

File: doc/external-access.md (hunk 4 above)
Comment (SrishT) [reply doc/external-access.md:243]: `For Exposing SegmentStoreservices on the same I/P address we will use MetalLB.` The line currently ends in a `,`

File: doc/external-access.md [hunk 9]
```diff
@@ -238,3 +238,31 @@ IP:                       10.100.200.183
...
+MetalLB hooks into Kubernetes cluster, and provides a network load-balancer implementation. In short, it allows to create Kubernetes services of type “LoadBalancer” in clusters that don’t run on a cloud provider and thus cannot simply hook into paid products to provide load-balancers.
+
+By default, Services do not share an IP address, for providing same IP address to all the services we need to set the following configurations while creating the External Service:
+
+1) Provide annotation key as metallb.universe.tf/allow-shared-ip for all the services.
```
Comment (SrishT) [inline doc/external-access.md:248]: Also the annotation could be highlighted like `metallb.universe.tf/allow-shared-ip`

File: charts/pravega/templates/pravega.yaml (hunk 3 above)
Comment (Prabhaker24) [reply charts/pravega/templates/pravega.yaml:36]: I tried doing that but I am getting this error ``` spec.pravega.segmentStoreLoadBalancerIP in body must be of type string: "null" ```

File: pkg/controller/pravega/pravega_segmentstore.go (hunk 1 above)
Comment (Prabhaker24) [reply pkg/controller/pravega/pravega_segmentstore.go]: Done

File: README.md (hunk 2 above)
Comment (Prabhaker24) [reply README.md]: Done

File: doc/external-access.md (hunk 4 above)
Comment (Prabhaker24) [reply doc/external-access.md]: Done.

File: charts/pravega/README.md (hunk 5 above)
Comment (Prabhaker24) [reply charts/pravega/README.md]: Done.

File: charts/pravega/templates/pravega.yaml (hunk 6 above)
Comment (Prabhaker24) [reply charts/pravega/templates/pravega.yaml]: No, I have changed it back.

File: charts/pravega/templates/pravega.yaml (hunk 7 above)
Comment (Prabhaker24) [reply charts/pravega/templates/pravega.yaml]: I have changed it back.

File: pkg/apis/pravega/v1beta1/pravega.go (hunk 8 above)
Comment (Prabhaker24) [reply pkg/apis/pravega/v1beta1/pravega.go]: Done

File: doc/external-access.md (hunk 9 above)
Comment (Prabhaker24) [reply doc/external-access.md]: Done.
